fixes #14661 - support IPv6 subnets

timogoebel opened the pull request:
This adds another Subnet type: IPv6!

The IPv6 class is based on Dominic's IPv6 Branch: https://github.com/theforeman/foreman/compare/develop...domcleal:2315-ipv6

timogoebel committed:
fixes #14661 - support IPv6 subnets

dLobatog commented on app/models/subnet.rb:
```diff
@@ -152,12 +152,16 @@ def cidr=(cidr)
 
   def dhcp?
-    !!(dhcp and dhcp.url and !dhcp.url.blank?)
+    !!(dhcp && dhcp.url && !dhcp.url.blank? && self.class.supports_dhcp?)
```
I think this expression should check for `supports_dhcp?` first, the rest doesn't make sense if that one is false. Also, checking for `.url` and `!url.blank?` is equivalent to `url.present?` (I know it's not your fault here but hey :smile: )

``` ruby
self.class.supports_ipam_mode?(:dhcp) && dhcp && dhcp.url.present?
```

I don't think the `!!` is needed here, it'll always be a boolean expression

dLobatog commented on app/models/subnet.rb:
```diff
@@ -152,12 +152,16 @@ def cidr=(cidr)
+    supported_ipam_modes.include?(mode)
+  end
+
+  def self.supports_dhcp?
```
I think you can do without this method and use `supports_ipam_mode?(:dhcp)` easily. Not only that, but also you can avoid calling `self.class` so often by using delegate.

``` ruby
delegate :supports_ipam_mode?, to: 'self.class'
```

That would allow you to call `supports_ipam_mode?(whatever)` everywhere in a cleaner way (in my opinion)

dLobatog commented on app/models/subnet.rb:
```diff
@@ -216,6 +220,12 @@ def unused_ip(mac = nil, excluded_ips = [])
       end
       logger.debug("Not suggesting IP Address for #{self} as no free IP found in our DB")
       return
+    elsif self.ipam == IPAM_MODES[:eui64] && self.class.supports_ipam_mode?(:eui64)
```
The whole `unused_ip` method should be implemented on `Subnet::Ipv4` and `Subnet::Ipv6`. What these `if` expressions are doing is basically class checking. 

`dhcp` and `db` should be in `unused_ip` `Subnet::Ipv4` and `eui64` should be in `unused_ip` in `Subnet::Ipv6`. There would be 3 lines of duplicated code at the beginning, lines 197 to 200 of this patch but that should be alright

timogoebel replied:
db will also be in ipv6, it had to remove it because it needs changes to the host or nic model...
I think, I could move that to a dedicated class or something. But the question is: Where to start and when to stop... I'll take a look and see if i can improve that without much effort...

dLobatog replied:
Changing it in this commit for the time being would drive the design to be a bit cleaner IMO, you would also be able to get rid of `self.supports_ipam_mode?(mode)` here if `unused_ip` is implemented in the subclasses (less code to maintain). I imagine the `db` part of it will be implemented as a service if it needs to be shared between both.

dLobatog commented on app/models/subnet.rb:
```diff
@@ -152,12 +152,16 @@ def cidr=(cidr)
 
-  def supports_ipam_mode?(mode)
-    self.class.supported_ipam_modes.include?(mode)
+  def self.supports_ipam_mode?(mode)
```
Nitpick, move all `self.` methods to `class << self` below the validations, this class is becoming a bit unwieldy

dLobatog commented on app/models/subnet.rb:
```diff
@@ -216,6 +220,12 @@ def unused_ip(mac = nil, excluded_ips = [])
+      return unless mac.present?
+      ip = mac_to_ip(mac)
+      logger.debug("Found #{ip}")
+      return ip
```
redundant return

dLobatog commented on app/models/subnet/ipv6.rb:
```diff
@@ -0,0 +1,50 @@
+      raise Foreman::Exception.new(N_("'%s' is not a valid MAC address.") % mac)
+    end
+    mac.gsub!(/[\.\:\-]/, '')
+    IPAddr.new(IPAddr.new(network, Socket::AF_INET6).to_i | ((mac.slice(0..5) + 'fffe' + mac.slice(6..11)).to_i(16) ^ 0x0200000000000000), Socket::AF_INET6).to_s
```
What is this doing? Can you split it in parts and leave a few comments?

dLobatog commented:
Thanks @timogoebel ! I left a few comments about the classes, they should be easy to fix. I haven't tried to use it or taken a close look at tests yet as code might change a bit. Let me know how it goes :smile:

timogoebel commented:
- moved `self.` to `class << self`
- moved `self.import` and `self.parse_dhcp_options` to Subnet::Ipv4 
- created a new ipam service class. I hope, you like it @dLobatog . I did not add a lot of tests because the classes are quite well covered by the subnet tests imho.

dLobatog commented on app/models/subnet/ipv4.rb:
```diff
@@ -35,6 +35,30 @@ def cleanup_addresses
+  end
+
+  # Translate ISC dhcp subnet options names provided by dhcp proxy into foreman subnet attributes names
+  def self.parse_dhcp_options(options)
```
This used to be a private method, I think it should continue to be one :sweat:

dLobatog commented on app/controllers/subnets_controller.rb:
```diff
@@ -58,7 +58,7 @@ def freeip
   def import
     proxy = SmartProxy.find(params[:smart_proxy_id])
-    @subnets = Subnet.import(proxy)
+    @subnets = Subnet::Ipv4.import(proxy)
```
We might want to show something on the UI specifying we're importing only IPv4 subnets (or is IPv6 import coming later)?

dLobatog replied:
Fixed by https://github.com/theforeman/foreman/pull/3442/files#diff-5b769784fd39cddb57cf503cd6f2392eR28

dLobatog commented on app/services/ipam/base.rb:
```diff
@@ -0,0 +1,23 @@
+    def normalize_mac!
+      return unless self.mac.present?
+      self.mac = Net::Validations.normalize_mac(self.mac)
+    rescue ArgumentError
```
I realize you're just dealing with a bad exception name from `Net::Validations`, but `ArgumentError` is supposed to be used when there [are not enough arguments, too many, etc..](http://ruby-doc.org/core-2.3.1/ArgumentError.html), not for exceptions having to do with the code. 

There's another instance of this on `Host::Base.get_interface_scope`  and `Nic::Base.normalize_mac`. Can you change `normalize_mac` to raise `Error` and these methods to rescue `Net::Validations::Error` instead? It'll only get more confusing if we let this bad exception name creep in.

dLobatog commented on app/models/subnet.rb:
```diff
@@ -195,28 +172,13 @@ def unused_ip(mac = nil, excluded_ips = [])
+
+    opts = {:subnet => self, :mac => mac, :excluded_ips => excluded_ips}
+    ipam = IPAM.new(self.ipam, opts)
+    ipam.suggest_ip
```
:clap:

dLobatog commented:
I made a few comments, nothing very major and we're definitely close to a final version with your latest changes :+1: . Sorry for the slow feedback lately, I'll try to pay more attention to updates so we can get this sufficiently ahead of 1.12. Thanks again @timogoebel

dLobatog commented on test/unit/subnet/ipv4_test.rb:
```diff
@@ -226,7 +226,7 @@ def create_a_domain_with_the_subnet
 
   test "should find unused IP on proxy if proxy is set" do
-    subnet = FactoryGirl.create(:subnet_ipv4, :name => 'my_subnet', :network => '192.168.1.0')
+    subnet = FactoryGirl.create(:subnet_ipv4, :ipam_dhcp, :name => 'my_subnet', :network => '192.168.1.0')
```
This and other tests down here are really testing the IPAM service class, they should be on a test file of it's own w/o subnet. The `Subnet` test should just test that it calls `suggest_ip`, that's all it ought to be concerned about.

timogoebel commented:
@dLobatog : Here you go! :-)
- replaced the ArgumentError with Net::Validations::Error. 👍 
- moved the ipam tests
- made self.parse_dhcp_options private using private_class_method
- made it clear in the ui that import only works for IPv4. I don't have any DHCPv6 code, yet.

dLobatog commented on app/models/subnet/ipv4.rb:
```diff
@@ -35,6 +35,32 @@ def cleanup_addresses
+    attrs
+  end
+
+  private_class_method :parse_dhcp_options
```
Why not just moving it below `private`?

iNecas replied:
Because private section does influence only instance methods

dLobatog replied:
Ah, then defining `class << self` might be a better option then. It's always confusing to have class methods & properties scattered

dLobatog commented on test/unit/ipam_test.rb:
```diff
@@ -0,0 +1,90 @@
+
+  context 'dhcp' do
+    test "should find unused IP on proxy if proxy is set" do
+      subnet = FactoryGirl.create(:subnet_ipv4, :ipam_dhcp, :name => 'my_subnet', :network => '192.168.1.0')
```
`.build`

dLobatog commented on test/unit/ipam_test.rb:
```diff
@@ -0,0 +1,90 @@
+
+  context 'internal db' do
+    test "should find unused IP in internal DB if proxy is not set" do
+      host = FactoryGirl.create(:host)
```
`.build` (just host)

dLobatog replied:
Actually I think this is testing too much. By creating hosts, we're testing `.known_ips` and stuff we shouldn't be testing here. [That's tested somewhere else already](https://github.com/theforeman/foreman/blob/develop/test/unit/subnet/ipv4_test.rb#L287). We just want to get a free IP from the db. So two options to offload this test are: 

1 - passing `:excluded_ips` to the IPAM service.
2 - stubbing `subnet.known_ips` to return whatever you want.

It'd look like this, you'd save creating an expensive object in db, and also it's much clearer. To understand the current test now one has to check the implementation.

``` ruby
test do
  subnet = FactoryGirl.create(
          :subnet_ipv4, :name => 'my_subnet',
          :network => '192.168.2.0',
          :ipam => IPAM::MODES[:db])
  ipam = IPAM::Db.new(:subnet => subnet, :excluded_ips => ['192.168.2.1', '192.168.2.2'])
  assert_equal '192.168.2.3', ipam.suggest_ip
end 
```

dLobatog commented on test/unit/ipam_test.rb:
```diff
@@ -0,0 +1,90 @@
+    end
+
+    test "should find unused IP excluding named values in internal DB if proxy is not set" do
+      host = FactoryGirl.create(:host)
```
Ditto https://github.com/theforeman/foreman/pull/3442/files#r61550858

dLobatog commented on test/unit/ipam_test.rb:
```diff
@@ -0,0 +1,90 @@
+    end
+
+    test "should respect subnet from and to if it's set" do
+      host = FactoryGirl.create(:host)
```
Ditto, no need to create hosts and nic_manageds https://github.com/theforeman/foreman/pull/3442/files#r61550858

dLobatog commented on test/unit/ipam_test.rb:
```diff
@@ -0,0 +1,90 @@
+    end
+
+    test 'should raise an exception if prefix length is not suitable' do
+      s = FactoryGirl.build(:subnet_ipv6, :network => '2001:db8::', :cidr => 70)
```
`subnet` reads better than `s`, the keyboard erosion you saved on this isn't worth the lack of readability :tongue:

dLobatog commented on test/unit/subnet/ipv4_test.rb:
```diff
@@ -225,68 +225,23 @@ def create_a_domain_with_the_subnet
-    subnet.reload
-    FactoryGirl.create(:nic_managed, :ip => '192.168.2.12', :subnet_id => subnet.id, :host => host, :mac => '00:00:00:00:00:03')
-    assert_nil subnet.unused_ip
+  test "#unused_ip should suggest IP" do
```
:clap:

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    User.current = User.find_by_login "admin"
+  end
+
+  test "should have a network" do
```
Use shoulda-matchers, `should validate_presence_of(:network)` does the same just by looking at the class validations instead of trying them out

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    assert @subnet.valid?
+  end
+
+  test "should have a mask" do
```
`should validate_presence_of(:mask)`

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    assert_equal 'ffff:ffff:ffff:ffff::', @subnet.mask
+  end
+
+  test "network should have ip format" do
```
`should_not allow_value("asf:fwe6::we6s:q1").for(:network)`

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    refute @subnet.valid?
+  end
+
+  test "mask should have ip format" do
```
`should_not allow_value("asf:fwe6::we6s:q1").for(:mask)`

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    refute @subnet.valid?
+  end
+
+  test "network should be unique" do
```
`should validate_uniqueness_of(:network)`

dLobatog replied:
Actually that comes from the parent class. so you should modify `test/unit/subnet.rb` to include these validations that are not specific from ipv6 to put them there. I saw `ipv4_test` also contains a bunch of these, just test them on the parent class for both. I missed that in the last pr, sorry.

You'll have to rename the file to `subnet_test.rb`, I don't think `/test/unit/subnet.rb` is actually being run now.

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    refute s.valid?
+  end
+
+  test "should strip whitespace before save" do
```
If I understand this correctly, this is testing `normalize_addresses`, that's on the parent class. It's also on https://github.com/theforeman/foreman/blob/develop/test/unit/subnet/ipv4_test.rb#L149 so it's basically a duplicate test.

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    assert_equal '2001:db8::3', s.dns_secondary
+  end
+
+  test "should not allow an address greater than 45 characters" do
```
`should_not allow_value(9.times.map { 'abcd' }.join(':')).for(:mask) # 45 characters`

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    refute FactoryGirl.build(:subnet_ipv6, :network => '2001:db8::1:').valid?
+  end
+
+  test "more than 4 chars in a block is invalid" do
```
`should_not allow_value('2001:db8:abcde::1').for(:network)`

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    assert_match /is invalid/, s.errors.full_messages.join("\n")
+  end
+
+  test "trailing colons in addresses are invalid" do
```
`should_not allow_value('2001:db8::1:').for(:network)`

dLobatog commented on test/unit/subnet/ipv6_test.rb:
```diff
@@ -0,0 +1,149 @@
+    refute FactoryGirl.build(:subnet_ipv6, :network => '2001:db8:abcde::1').valid?
+  end
+
+  # test module StripWhitespace which strips leading and trailing whitespace on :name field
```
Again, this is tested on ipv4, so move them both to the test for the parent class (it's where these things are happening)

timogoebel commented:
@dLobatog : I cleaned up the tests. I hope, I did not miss anything. Cloud you double check, please?

dLobatog commented:
@timogoebel I think this is pretty solid already. Thanks for the updates :smile: ! I'll find some time for testing, probably on Tuesday and hopefully we can merge it then.

dLobatog commented on app/services/ipam.rb:
```diff
@@ -0,0 +1,16 @@
+module IPAM
+  MODES = {:dhcp => N_('DHCP'), :db => N_('Internal DB'), :eui64 => N_('EUI-64'), :none => N_('None')}
```
Since now this supports eui-64, can you add a short description to this list? https://github.com/theforeman/foreman/blob/develop/app/views/subnets/_fields.html.erb#L13 

It currently shows up as an option available, but there's no help for it

dLobatog replied:
Scratch that, actually. There are a few other things that would need to be changed in order for this to work (like auto-detecting the IP type based on how it looks), so let's do that in the next PR concerning UI. I'll just test it manually creating stuff in the console

timogoebel commented:
@dLobatog : How did testing go? Did you reach a final verdict?

dLobatog commented:
Took me a while to set up everything in the console but IMO all ready, eui-64 working well and everything. Just 1 last question, why is :db not in the list of supported IPAM modes for ipv6? (edit: responded on IRC)

timogoebel commented:
... just for reference: it needs changes to the host model, PR incoming...

@dLobatog : Thanks a lot for reviewing this. I really like this last version.

dLobatog commented:
Merged as 990dee696f58b9fa6035b239152223b073999488, thanks @timogoebel!
